src/Model/ChangeStreamIterator.php

```diff
@@ -143,6 +135,14 @@ public function current()
         return $this->isValid ? parent::current() : null;
     }
 
+    final public function getInnerIterator(): Cursor
```

Comment:
Adding this method to reduce the number of assertions needed in the `ChangeStream` model class. Strictly speaking any instance of `ChangeStreamIterator` would return a cursor when calling `getInnerIterator()`, as that's enforced through the constructor. However, as the class isn't final and there are no covariance checks for constructors, a user could potentially sidestep this.

The `assert` call is to satisfy psalm, and when assertion checking is disabled a type error will be thrown. I believe this is better than letting a user run into an error about calling a method that doesn't exist. Let me know if you'd rather remove this method and instead use assertions in `ChangeStream`.

Note that this method can be removed when making the class final, which we should do in 2.0.

Reply:
Doesn't this only affect `ChangeStream::getCursorId()`? I don't see any other place where `getInnerIterator()` is called in the library apart from two instances in `WatchFunctionalTest`.

This is also interesting because ChangeStreamIterator extends IteratorIterator and enforces that its constructor is always a Cursor instance. I suppose psalm cannot infer that `IteratorIterator::getInnerIterator()` will always return an instance of _its_ constructor argument without additional templating.

No objection to leaving this method in place, but I think a doc block above that briefly explains why it's needed and that it can potentially be removed in 2.0 would be helpful. You can also make a ticket with a 2.0 fixVersion to remind us to remove this down the line.

Reply:
Yes and no. While psalm 

> Doesn't this only affect `ChangeStream::getCursorId()`? I don't see any other place where `getInnerIterator()` is called in the library apart from two instances in `WatchFunctionalTest`.

That is correct.

> This is also interesting because ChangeStreamIterator extends IteratorIterator and enforces that its constructor is always a Cursor instance. I suppose psalm cannot infer that `IteratorIterator::getInnerIterator()` will always return an instance of _its_ constructor argument without additional templating.

Yes and no. While this is certainly true within the context of `ChangeStreamIterator`, technically an extending class could skip calling the parent constructor and go straight to `IteratorIterator::__construct`: https://3v4l.org/iir32. In that case, our assumption that the inner iterator is always a cursor no longer holds true. Yes, it is very much theoretical, but then that's what psalm was made for.

> No objection to leaving this method in place, but I think a doc block above that briefly explains why it's needed and that it can potentially be removed in 2.0 would be helpful. You can also make a ticket with a 2.0 fixVersion to remind us to remove this down the line.

Added 👍